### DISCUSSION TITLE
Provisioning devices and getting them onto end-user WiFi networks

### DISCUSSION THREAD

**Discussion by justind000** [toitlang/toit#80] [Ideas]
I have been looking at how to provision a Toit device. I am looking from the perspective of an end-user who has received a device that has been programmed with Toit and now has to set it up to work at their location.

How does the end-user enter their WiFi settings at this point? The end-user shouldn't need to interact through Toit's console or CLI as they bought a device and just want it to work with a simple setup process. 

Here a some ideas to accomplish this:

- The product provider could have a website which gives instructions to connect the device to a computer using a USB cable then goto a website (in Chrome, Edge or Opera for [Web Serial support](https://caniuse.com/web-serial)). That website has the end-user enter setup information. This method could be used in native apps as well, removing the browser limitation. 
- Espressif offers several [provisioning ](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_smartconfig.html)options in the IDF. 
    - Smartconfig - Press a button on the WiFi router while putting the ESP32 in a mode to receive the WiFi information. This is quite easy, but not everyone has a Smartconfig-capable router. 
    - Easy Connect - Scan QR codes on both router and Toit device. Similarly to Smartconfig, requires Easy connect-enabled router. 
    - [Unified Provisioning](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/provisioning/provisioning.html): Uses either WiFi (self-hosted access point), BLE, or serial interface to connect and exchange information in protobuf format. All three methods require a dedicated website or app to handle the communication and user-interaction. The WiFi access point method is troublesome as most devices don't like to be connected to a WiFi network that doesn't provide internet access and will complain/disconnect. 
 - Outside repos that implement the access-point and BLE method exist as well. Some also provide a website that is device-hosted for the user to provide WiFi information. 

I have tried most of the methods and they all worked. Espressif provides [sample apps ](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/provisioning/provisioning.html#provisioning-tools)to use the Unified Provisioning and they work. I think the most reliable method is the BLE version. As mentioned above, the access point method (both Espressif and outside repos) creates a WiFi network for the user to connect to. My phone complains about it not having internet access, as does my computer. I need to tap/click a few times to tell it to stay connected. 

I was hoping to get a conversation going regarding other methods or ideas, experiences etc. Also to see if there is any interest on the Toit side to implement a particular method for this use-case, as any method will require additional code, but also some rewrites to the WiFi connection logic (what to do if no WiFi).

**Comment by kasperl** [comment]
Hi @justind000! We're definitely interested in exploring this with you. Should we try to figure out how to make the BLE based WiFi configuring work? I believe we have most of the building blocks.

**Reply by justind000** [reply]
Speaking for my personal intentions, I would like a BLE-based provisioning interface. But I believe using the provided Unified service, the access point version is very close to identical in terms of firmware code. 

I'll need to get familiarized with the IDF [provisioning example](https://github.com/espressif/esp-idf/blob/4143a3f572f6060258ffac91f110a095a8478966/examples/provisioning/wifi_prov_mgr/main/app_main.c). A quick glance has brought up a few places that could use some discussion:
- If a device has no WiFi connection, should it automatically run the provisioning service or should it be a more programmatic process? Automatic would be convenient, but there are situations it may be unwanted: the configured router may just be temporarily unavailable or the device might have been temporarily moved being the first two I thought of. 
  In the past I have provided the end-user access to a configuration process through pressing a button. They can then enter multiple WiFi networks for the device to try and connect to. On the next device boot, it will try to connect, if it cannot the device will goto deep sleep and try again at the next reporting interval. This works for my purposes, but I wouldn't presuppose everyone wants this behavior. 
   How does connection status bubble up to the programmer? I haven't been able to find how to access it from the Toit VM. Maybe running the Toit code in its own [application ](https://docs.toit.io/platform/apps/appspec) that runs on_boot would work well. 

- An app will need to accompany this. There are limitless ways to accomplish this, native Android/iOS apps, [Web Bluetooth](https://caniuse.com/web-bluetooth) through compatible browsers or something like [Capacitor](https://capacitorjs.com/) which uses HTML to generate native-like apps. The provisioning service could be used to pass other configuration information or interact with the device in other ways, so the app could be a simple provisioning only process, or be a small part of a larger application. 
   I can make a Web Bluetooth and Capacitor version, but I'm no expert by any stretch and know nearly nothing about native app development. Perhaps a variety of demos will eventually be available as a starting point.

**Reply by justind000** [reply]
Just a quick update. I have gotten the [sample BLE app](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/provisioning/provisioning.html#provisioning-tools) to connect and pass credentials to Toit. 

I ended up with this:
```
main:
  print "starting provisioning"
  #primitive.provision.init
```
Which sets up a BLE server you can connect to and interact with. More to work on of course, but that was just a PoC for me to get familiar with things. 

I will start to work on a Web Bluetooth-based app in the coming days.

**Reply by lazyoldjack** [reply]
I too was considering how this would work for the best, in a simple reliable manner that enabled the least technical end-user to connect his purchase (the widget) to his Wi-Fi network. For me, the assumption is that whoever has purchased the widget can manage the basic use of a phone or computer  so can (preferably) make a Bluetooth connection or otherwise can connect to a default AP presented by the ESP32. Does Toit remove that AP on 192.168.4.1 functionality? I haven't looked for it or tried it...

Previously, in MicroPython, I have presented a web page on the AP (when no wifi connection was achievable) that allows the end-user to enter their wifi credentials and then it's all good to go. While this does have the advantage of not having to install anything on the phone or computer, it's not very slick; I would be interested in anything better that "just works"!

One thing that did occur to me was that use of an SSID of a particular format for the wifi network used in testing the widget before despatch to the end-user could inform the Toit application that this is a wifi that it should not waste its time looking for if it's not there, so that the "end-user routine" could kick in.

Sorry - I don't seem to have added much to the conversation apart from agreeing that something slick, reliable and simple for the end-user is required.

**Reply by justind000** [reply]
@lazyoldjack I have implemented the Espressif provisioning functionality. It allows you to create either an AP or BLE connection from which you can use [Espressif provisioning mobile apps](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/provisioning/provisioning.html#provisioning-tools) to connect and provide user credentials. The source for the apps is provided as well, so making your own device specific app would be possible, with the provisioning being a part of what the app does. 

It is here, https://github.com/justind000/toit/commit/937cdee564d889e70a75c84815a874629cfecf66, however there were a couple things that the Toit folks needed to work on prior to it getting pulled in. 

Right now it would work like this:
```
import provision
main:
  print "starting provisioning"
  provision := provision.AP "PROV_ap"
  //provision := Provision.BLE "PROV_ble"
  ret := provision.start
  print (ret ? "Provisioning success" : "Provisioning failed")
```

**Reply by dashcraft** [reply]
I'd like a standard similar to:
https://www.improv-wifi.com/

Which is straightforward and open source

**Reply by floitsch** [reply]
Agreed. I don't think we looked at this particular library, but we had something similar in mind.
The biggest problem is that you still need some other approach as iOS devices don't support it.

**Reply by justind000** [reply]
That's very similar to the Espressif system discussed prior in this thread. 

There is a link to an (at this point old) working implementation of it with links to the Espressif apps as well.

**Reply by dashcraft** [reply]
I have a react native application that implements improv-wifi for android and ios, i can make it open source and publish it if we can get this in.

**Reply by kasperl** [reply]
We have added support for establishing a WiFi in Soft AP mode on the ESP32. Together with some simple DNS server code, you can use this to create a provisioning flow.

See https://github.com/toitlang/pkg-dns-simple-server/blob/10476712e5140d40b9f593530aca50045ba41671/examples/captive_portal.toit#L65. Once you have the new WiFi credentials through some sort of form post, you can install them by doing a call to `wifi.open --save --ssid="bestwifi" --password="donttell"`.

/cc @justind000

**Comment by kasperl** [comment]
We have added support for establishing a WiFi in Soft AP mode on the ESP32. Together with some simple DNS server code, you can use this to create a provisioning flow.

See https://github.com/toitlang/pkg-dns-simple-server/blob/10476712e5140d40b9f593530aca50045ba41671/examples/captive_portal.toit#L65. 

Once you have the new WiFi credentials through some sort of form post, you can install them by doing a call to 

```
wifi.open --save --ssid="bestwifi" --password="donttell"`
```

This will save the credentials if we can successfully connect to the network and use the credentials for future attempts to connect to the network.

(I originally posted this as a reply to a comment, but I decided that it's meaningful to make this a separate comment)

**Reply by dashcraft** [reply]
So to be clear, you can take in the credentials via a form post (or any other methods) and then save them via code now?

**Reply by kasperl** [reply]
Yes, @dashcraft - finally :) 

The captive portal example - https://github.com/toitlang/pkg-dns-simple-server/blob/main/examples/captive_portal.toit - shows the hardest part which is establishing a WiFi network and answering DNS queries.

**Reply by dashcraft** [reply]
Nice, so this opens up the ability to do ble onboarding of wifi credentials as well using wifi.open --save --ssid="bestwifi" --password="donttell"`

**Reply by kasperl** [reply]
Indeed!

**Reply by ojmason** [reply]
Will this persist if the ESP32 device is powered down?

**Reply by erikcorry** [reply]
Yes, that's what the --save option does.

**Comment by kasperl** [comment]
The sample project https://github.com/kasperl/toit-zygote shows how to pull all this together.
